[Analytics Hub] Refactor Time Range data structure

ThomazFB opened the pull request:
Part of #8146 

#### Mostly repositioning and renaming class names and groups, don't be scared of the number of lines changed. 😅  

Why
==========
The previous `AnalyticsHubTimeRangeGenerator` class would be hard to maintain as we need to introduce new range selections like `last month`, `last year`, and new types like Quarter ranges. We need a simple way to define those ranges as they have a pattern repetition, configuring those ranges for reuse but keeping the current API.

How
==========
This PR introduces dedicated time range classes responsible for building the main ranges based on a reference date. The reference date can be changed to support different selection types, allowing those classes to be reused to generate types like `last month`, `yesterday, `last week` and `last year`. 

The `AnalyticsHubViewModel` will continue to only refer to the `AnalyticsHubTimeRangeSelection`, and that class has a delegation system that will call incorporate the correct range class based on the selected range type.

How to Test
==========
1. Launch the app and click the `See More button in the My Store charts section.
2. Verify that the date ranges and range selections are working as expected.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Define AnalyticsHubTimeRangeSelection protocol

ThomazFB committed:
Remove AnalyticsHubTimeRange struct from the AnalyticsHubTimeRangeSelectionData

ThomazFB committed:
Define AnalyticsHubMonthRangeSelection

ThomazFB committed:
Define AnalyticsHubYearRangeSelection

ThomazFB committed:
Define AnalyticsHubWeekRangeSelection

ThomazFB committed:
Define AnalyticsHubDayRangeSelection

ThomazFB committed:
Refactor AnalyticsHubTimeRangeSelectionData to take advantage of the individual implementation of ranges

ThomazFB committed:
Simplify TimeRange description generation

ThomazFB committed:
Adjust description declaration to the inside of the AnalyticsHubTimeRange

ThomazFB committed:
Adopt full usage of the TimeRangeSelection inside AnalyticsHubTimeRangeSelectionData

ThomazFB committed:
Configure internal description inside the TimeRange Selection

ThomazFB committed:
Refactor AnalyticsHubTimeRangeSelectionData properties

ThomazFB committed:
Fix AnalyticsHubDayRangeSelection description

ThomazFB committed:
Make AnalyticsHubTimeRangeSelectionData selection description public

ThomazFB committed:
Rename TimeRange group of classes

ThomazFB committed:
Merge branch 'issue/8146-introduce-time-range-picker' into issue/8146-introduce-last-range-structure

ThomazFB committed:
Fix TimeRange group of classes after merge

ThomazFB committed:
Revert SelectionType position

ThomazFB committed:
Update timezone usage to rely on the Calendar timezone

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubTimeRangeSelection.swift:
```diff
@@ -0,0 +1,138 @@
+    var previousRangeDescription: String? { get }
+
+    init(referenceDate: Date, currentCalendar: Calendar)
+}
```
I'm unsure about `Delegate` naming here. In iOS SDK delegates are usually active objects that perform some operations in place of original owner. Like when tableview delegate handles row selection, drag and other actions.

Here we have object purely for storage. It transforms init input into output and that's all. I'm inclined to call this one `AnalyticsHubTimeRangeSelection` and existing `AnalyticsHubTimeRangeSelection` -> `...SelectionFactory`.

It's also a bit confusing in the class hierarchy:
AnalyticsHub**Year**RangeSelection is a child of AnalyticsHub**Time**RangeSelection**Delegate**.
If the parent is a delegate, its child should be delegate too? 🤔

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubTimeRangeSelection.swift:
```diff
@@ -0,0 +1,138 @@
+            }
+        }
+
+        static func from(_ statsTimeRange: StatsTimeRangeV4) -> SelectionType {
```
Can we write it as a custom initializer (`init(_ statsTimeRange: StatsTimeRangeV4)`)? It better conveys the intent.

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8267/commits/3a70c3a3b3f29cad7fa2e2f0091413929baa5db3

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubTimeRangeSelection.swift:
```diff
@@ -0,0 +1,138 @@
+        case yearToDate
+
+        var description: String {
+            get {
```
nit: can be removed

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8267/commits/e456e6d45f2731ebd5fdfd459b1ecd11215a5a78

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubDayRangeSelection.swift:
```diff
@@ -0,0 +1,27 @@
+        } else {
+            self.previousTimeRange = nil
+            self.previousRangeDescription = nil
+        }
```
Q: is there a case to have correct `currentTimeRange`, but `previousTimeRange = nil`?

My idea is: if partial fail means inability to display relevant information to the user, we can make RangeSelection initializers failable. So if something goes wrong in init phase - fail and throw readable error instead of creating a data object with all values = nil.

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8267/commits/1077d07c3186a793daffece808c0132f1cc787cb

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubDayRangeSelection.swift:
```diff
@@ -0,0 +1,27 @@
+    var currentTimeRange: AnalyticsHubTimeRange?
+    var previousTimeRange: AnalyticsHubTimeRange?
+    var currentRangeDescription: String?
+    var previousRangeDescription: String?
```
Can we replace all `var`s in all Selection objects with `let`s?

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8267/commits/1077d07c3186a793daffece808c0132f1cc787cb

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubDayRangeSelection.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+final class AnalyticsHubDayRangeSelection: AnalyticsHubTimeRangeSelectionDelegate {
```
Can it be a `struct`?

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8267/commits/1077d07c3186a793daffece808c0132f1cc787cb

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubTimeRangeSelection.swift:
```diff
@@ -0,0 +1,138 @@
+    /// Generates a date description of the previous time range set internally.
+    /// - Returns the Time range in a UI friendly format. If the previous time range is not available,
+    /// then returns an presentable error message.
+    func generateCurrentRangeDescription() -> String {
```
Since it just provides convenient access for the variable - can we make it computed `var currentRangeDescription` instead of a `func`?
From `func` API I expect some processing inside or additional side effects.

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8267/commits/5f4ad85a17f30cee5f93f6cb580dab4b1844e064

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubTimeRangeSelection.swift:
```diff
@@ -0,0 +1,138 @@
+    /// Generates a date description of the previous time range set internally.
+    /// - Returns the Time range in a UI friendly format. If the previous time range is not available,
+    /// then returns an presentable error message.
+    func generatePreviousRangeDescription() -> String {
```
Same as above

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8267/commits/5f4ad85a17f30cee5f93f6cb580dab4b1844e064

ealeksandrov reviewed (commented):
Thanks for focusing on architecture and code readability!
I've left some ideas and questions around the code - they may be already discussed in previous PRs and some are just style/naming choices, so please bear with me 😅

Generally I'm a bit concerned that all `AnalyticsHub...RangeSelection` objects differ only in 2-4 lines out of 30. They have different inputs to `AnalyticsHubTimeRange(start: ..., end: ...)`, but the same logic otherwise. 4 properties have the same types, but still declared separately in each range implementation. Can we add a superclass holding shared logic?

As I understand, we will have more complexity adding next set of ranges later, so current implementation may fit better there. We can keep the architecture and discuss/reiterate later! Let me know what you think!

ThomazFB committed:
Overhaul TimeRange generation internal structure

ThomazFB committed:
Introduce AnalyticsHubTimeRangeData protocol file

ThomazFB committed:
Cover simplified time range description scenario

ThomazFB committed:
Fix round issues

ThomazFB committed:
Introduce timezone parameter

ThomazFB committed:
Improve generateDescription parameter usage

ThomazFB committed:
Use a init for the SelectionType

ThomazFB committed:
Rename timeRangeSelection field naming

ThomazFB committed:
Remove unnecessary get from SelectionType.description

ThomazFB committed:
Refactor description generation functions to computed properties

ThomazFB committed:
Fix AnalyticsHubTimeRangeSelectionTests

ThomazFB commented:
@ealeksandrov thank you for this great review 🙇 Your inputs made me rethink my approach here, and I think I managed to simplify the time range structure while keeping my objective of making the implementation easier of the upcoming additional ranges. 

I've removed this "Delegate" idea, moved to data classes responsible for building only the specifics of each range, and added common code behavior as a protocol extension. I believe this solves most of the code repetition while leaving room for reuse. 

Let me know what you think about those changes! I'm still unsure about the naming pattern I've adopted, so let me know what you think of this current implementation. 

I've also addressed the small adjustments you suggested. The PR is ready for a second round!

ealeksandrov reviewed (approved):
LGTM! :shipit:
